### PR TITLE
Add makefile target 'show_build_options' and 'show_all_features'

### PR DESCRIPTION
**Pull request by mtei**
## Description

Example:
```
$ make  planck:default:show_all_features
QMK Firmware 0.6.266
Making planck/rev5 with keymap default and target show_all_features

 PLATFORM = AVR
 FIRMWARE_FORMAT = hex

Build Options:
  BOOTMAGIC_ENABLE = no
  MOUSEKEY_ENABLE = no
  EXTRAKEY_ENABLE = yes
  CONSOLE_ENABLE = yes
  COMMAND_ENABLE = no
  SLEEP_LED_ENABLE = no
  NKRO_ENABLE = yes
  BACKLIGHT_ENABLE = no
  RGBLIGHT_ENABLE = no
  MIDI_ENABLE = no
  UNICODE_ENABLE = no
  BLUETOOTH_ENABLE = no
  AUDIO_ENABLE = yes
  FAUXCLICKY_ENABLE =
  HD44780_ENABLE =

Other Features:
  API_SYSEX_ENABLE = no
  RGB_MATRIX_ENABLE = no
```

## Types of changes

- [x] Core
- [ ] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: show_options.mk [hunk 1]
```diff
@@ -0,0 +1,115 @@
+show_build_options:
+	echo " PLATFORM = $(PLATFORM)"
```
Comment (vomindoraan) [inline show_options.mk]: Inside recipes (lines starting with a tab character), shell syntax is used, so `$()` creates a subshell. That's not what you want here; you want parameter substitution (aka. interpolation). Use `$var` or `${var}` instead.
```suggestion
	echo " PLATFORM = $PLATFORM"
```
Make this change to all the lines below as well.

File: show_options.mk [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	echo " FIRMWARE_FORMAT = $(FIRMWARE_FORMAT)"
+	echo
+	echo Build Options:
```
Comment (vomindoraan) [inline show_options.mk]:
```suggestion
	echo "Build Options:"
```
Add missing quotes.

File: show_options.mk [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	echo "  UNICODE_ENABLE = $(UNICODE_ENABLE)"
+	echo "  BLUETOOTH_ENABLE = $(BLUETOOTH_ENABLE)"
+	echo "  AUDIO_ENABLE = $(AUDIO_ENABLE)"
+	echo "  FAUXCLICKY_ENABLE = $(FAUXCLICKY_ENABLE)"
+	echo "  HD44780_ENABLE = $(HD44780_ENABLE)"
+
+
+show_all_features: show_build_options
+	echo
+	echo Other Features:
```
Comment (vomindoraan) [inline show_options.mk]:
```suggestion
	echo "Other Features:"
```

File: show_options.mk [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    ifneq ($(API_SYSEX_ENABLE),)
+	echo "  API_SYSEX_ENABLE = $(API_SYSEX_ENABLE)"
```
Comment (vomindoraan) [inline show_options.mk]:
```suggestion
	echo "  API_SYSEX_ENABLE = $API_SYSEX_ENABLE"
```
Use parameter substitution here and on all `echo` lines below.

File: show_options.mk [hunk 5]
```diff
@@ -0,0 +1,115 @@
+show_build_options:
```
[the remark this answers is not in the document]
Comment (mtei) [reply show_options.mk]: Thanks.

File: show_options.mk [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	echo "  BACKLIGHT_ENABLE = $(BACKLIGHT_ENABLE)"
+	echo "  RGBLIGHT_ENABLE = $(RGBLIGHT_ENABLE)"
+	echo "  MIDI_ENABLE = $(MIDI_ENABLE)"
+	echo "  UNICODE_ENABLE = $(UNICODE_ENABLE)"
+	echo "  BLUETOOTH_ENABLE = $(BLUETOOTH_ENABLE)"
+	echo "  AUDIO_ENABLE = $(AUDIO_ENABLE)"
+	echo "  FAUXCLICKY_ENABLE = $(FAUXCLICKY_ENABLE)"
+	echo "  HD44780_ENABLE = $(HD44780_ENABLE)"
+
+
```
[the remark this answers is not in the document]
Comment (mtei) [reply show_options.mk]: Thanks.

File: show_options.mk (hunk 2 above)
Comment (mtei) [reply show_options.mk]: Thanks.

File: show_options.mk (hunk 3 above)
Comment (mtei) [reply show_options.mk]: Thanks.

**Comment by drashna** [comment]
This looks good.

But I see a number of issues with the separation here.

Like "UNICODE_ENABLE" is in the build options, but the other two unicode methods are listed under "other features".
Or RGBLIGHT_ENABLE is listed under build options, but RGB_MATRIX_ENABLE is listed under "other".

This creates a disconnect of what should be where. 


I think it may be a good idea to break this into three categories, actually: 
* Build Options
* Hardware Options
* Others

With Build containing:
BOOTMAGIC_ENABLE 
MOUSEKEY_ENABLE 
EXTRAKEY_ENABLE
CONSOLE_ENABLE 
COMMAND_ENABLE 
NKRO_ENABLE 
TERMINAL_ENABLE
CUSTOM_MATRIX
SPLIT_KEYBOARD
DYNAMIC_KEYMAP_ENABLE
TERMINAL_ENABLE
USB_HID_ENABLE

Hardware containing: 
SLEEP_LED_ENABLE 
BACKLIGHT_ENABLE 
BACKLIGHT_CUSTOM_DRIVER 
RGBLIGHT_ENABLE 
RGB_MATRIX_ENABLE
SERIAL_LINK_ENABLE
CIE1931_CURVE
MIDI_ENABLE 
BLUETOOTH_ENABLE 
AUDIO_ENABLE 
FAUXCLICKY_ENABLE 
HD44780_ENABLE 
ENCODER_ENABLE
LCD_ENABLE
LED_BREATHING_TABLE
LED_TABLES
POINTING_DEVICE_ENABLE
VISUALIZER_ENABLE

And all the rest in Other.

**Comment by drashna** [comment]
Awesome!  Thanks! 

And actually, that looks a lot more maintainable too!

**Comment by drashna** [comment]
Is there a way to sort the settings?

**Comment by mtei** [comment]
> Is there a way to sort the settings?

Sorting? Do you want to sort alphabetically?

**Comment by drashna** [comment]
I think that ideally, yeah, it should be alphabetical.

**Comment by drashna** [comment]
Awesome, thanks! 

Though, one last thing.... I noticed that a lot of the options are blank.  Would it be possible to set blank settings to "no" or "undefined" or something, so it's not just blank?

If not, then that's fine.

**Comment by mtei** [comment]
Example No.1:

```
$ make helix:default:show_build_options
QMK Firmware 0.6.266
Making helix/rev2 with keymap default and target show_build_options

 KEYBOARD = helix/rev2
 KEYMAP   = default
 PLATFORM = AVR
 FIRMWARE_FORMAT = hex

Build Options:
  BOOTMAGIC_ENABLE = no       		# file
  COMMAND_ENABLE = no        		# file
  CONSOLE_ENABLE = no         		# file
  CUSTOM_DEBOUNCE = 		# undefined
  CUSTOM_MATRIX = yes		# file
  DYNAMIC_KEYMAP_ENABLE = 		# undefined
  EXTRAKEY_ENABLE = no       		# file
  MOUSEKEY_ENABLE = no       		# file
  NKRO_ENABLE = no            		# file
  SPLIT_KEYBOARD = 		# undefined
  TERMINAL_ENABLE = 		# undefined
  USB_HID_ENABLE = 		# undefined

If you want to know more, please try 'show_all_features' or 'show_full_features'
```

**Comment by mtei** [comment]
Example No.2:

```
$ make RGBLIGHT_ENABLE=yes helix:default:show_all_features

QMK Firmware 0.6.266
Making helix/rev2 with keymap default and target show_all_features

 KEYBOARD = helix/rev2
 KEYMAP   = default
 PLATFORM = AVR
 FIRMWARE_FORMAT = hex

Build Options:
  BOOTMAGIC_ENABLE = no       		# file
  COMMAND_ENABLE = no        		# file
  CONSOLE_ENABLE = no         		# file
  CUSTOM_DEBOUNCE = 		# undefined
  CUSTOM_MATRIX = yes		# file
  DYNAMIC_KEYMAP_ENABLE = 		# undefined
  EXTRAKEY_ENABLE = no       		# file
  MOUSEKEY_ENABLE = no       		# file
  NKRO_ENABLE = no            		# file
  SPLIT_KEYBOARD = 		# undefined
  TERMINAL_ENABLE = 		# undefined
  USB_HID_ENABLE = 		# undefined

Hardware Options:
  AUDIO_ENABLE = no           		# file
  BACKLIGHT_ENABLE = no      		# file
  BLUETOOTH_ENABLE = no       		# file
  CIE1931_CURVE = yes		# file
  LED_BREATHING_TABLE = yes		# file
  LED_TABLES = yes		# file
  MIDI_ENABLE = no            		# file
  RGBLIGHT_ENABLE = yes		# command line
  RGB_MATRIX_ENABLE = no		# file
  SLEEP_LED_ENABLE = no    		# file

Other Options:
  IOS_DEVICE_ENABLE = no      		# file
  LED_ANIMATIONS = yes        		# file
  LED_BACK_ENABLE = no        		# file
  LED_UNDERGLOW_ENABLE = no   		# file
  OLED_ENABLE = no            		# file
  SWAP_HANDS_ENABLE = no        		# file
  UNICODE_ENABLE = no         		# file

```

```
$ make HELIX=back,oled,ios helix:default:show_all_features 

QMK Firmware 0.6.266
Making helix/rev2 with keymap default and target show_all_features

Helix customize
-  OLED_ENABLE=yes
-  LED_BACK_ENABLE=yes
-  LED_UNDERGLOW_ENABLE=no   
-  LED_ANIMATION=yes        
-  IOS_DEVICE_ENABLE=yes

 KEYBOARD = helix/rev2
 KEYMAP   = default
 PLATFORM = AVR
 FIRMWARE_FORMAT = hex

Build Options:
  BOOTMAGIC_ENABLE = no       		# file
  COMMAND_ENABLE = no        		# file
  CONSOLE_ENABLE = no         		# file
  CUSTOM_DEBOUNCE = 		# undefined
  CUSTOM_MATRIX = yes		# file
  DYNAMIC_KEYMAP_ENABLE = 		# undefined
  EXTRAKEY_ENABLE = no       		# file
  MOUSEKEY_ENABLE = no       		# file
  NKRO_ENABLE = no            		# file
  SPLIT_KEYBOARD = 		# undefined
  TERMINAL_ENABLE = 		# undefined
  USB_HID_ENABLE = 		# undefined

Hardware Options:
  AUDIO_ENABLE = no           		# file
  BACKLIGHT_ENABLE = no      		# file
  BLUETOOTH_ENABLE = no       		# file
  CIE1931_CURVE = yes		# file
  LED_BREATHING_TABLE = yes		# file
  LED_TABLES = yes		# file
  MIDI_ENABLE = no            		# file
  RGBLIGHT_ENABLE = yes		# file
  RGB_MATRIX_ENABLE = no		# file
  SLEEP_LED_ENABLE = no    		# file

Other Options:
  HELIX = back,oled,ios		# command line
  IOS_DEVICE_ENABLE = yes		# file
  LED_ANIMATIONS = yes        		# file
  LED_BACK_ENABLE = yes		# file
  LED_UNDERGLOW_ENABLE = no   		# file
  OLED_ENABLE = yes		# file
  SWAP_HANDS_ENABLE = no        		# file
  UNICODE_ENABLE = no         		# file

```

**Pull request merged**